workflow/util/util.go

```diff
@@ -735,36 +735,60 @@ func convertNodeID(newWf *wfv1.Workflow, regex *regexp.Regexp, oldNodeID string,
 }
 
 // RetryWorkflow updates a workflow, deleting all failed steps as well as the onExit node (and children)
-func RetryWorkflow(ctx context.Context, kubeClient kubernetes.Interface, hydrator hydrator.Interface, wfClient v1alpha1.WorkflowInterface, name string, restartSuccessful bool, nodeFieldSelector string) (*wfv1.Workflow, error) {
+func RetryWorkflow(ctx context.Context, podIf v1.PodInterface, wfClient v1alpha1.WorkflowInterface, name string, restartSuccessful bool, nodeFieldSelector string) (*wfv1.Workflow, error) {
+	var updatedWf *wfv1.Workflow
+	err := waitutil.Backoff(retry.DefaultRetry, func() (bool, error) {
+		var err error
+		wf, err := wfClient.Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		updatedWf, err = prepareWorkflowForRetry(ctx, wf, podIf, restartSuccessful, nodeFieldSelector)
+		if err != nil {
+			return false, err
+		}
+		updatedWf, err = wfClient.Update(ctx, updatedWf, metav1.UpdateOptions{})
```

Comment:
could you make this an upsert? then it would work for both archived and normal workflows, and would also be idempotent and therefore robust